File: src/main/java/com/atsuishio/superbwarfare/client/sound/ModSoundInstances.java
package com.atsuishio.superbwarfare.client.sound;

import com.atsuishio.superbwarfare.compat.netmusic.NetMusicCompatHolder;
import com.atsuishio.superbwarfare.entity.projectile.FastThrowableProjectile;
import com.atsuishio.superbwarfare.entity.vehicle.base.VehicleEntity;
import net.minecraft.client.Minecraft;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.api.distmarker.OnlyIn;

@OnlyIn(Dist.CLIENT)
public class ModSoundInstances {

    public static void init() {
        VehicleEntity.trackSound = vehicle -> Minecraft.getInstance().getSoundManager().play(new VehicleSoundInstance.TrackSound(vehicle));
        VehicleEntity.engineSound = vehicle -> Minecraft.getInstance().getSoundManager().play(new VehicleSoundInstance.EngineSound(vehicle));
        VehicleEntity.swimSound = vehicle -> Minecraft.getInstance().getSoundManager().play(new VehicleSoundInstance.SwimSound(vehicle));
        VehicleEntity.hornSound = vehicle -> Minecraft.getInstance().getSoundManager().play(new HornSoundInstance.VehicleHornSound(vehicle));
        VehicleEntity.inCarMusic = vehicle -> {
            if (NetMusicCompatHolder.canPlayMusic(vehicle)) {
                NetMusicCompatHolder.playMusic(vehicle);
            } else {
                Minecraft.getInstance().getSoundManager().play(new InCarMusicInstance.InCarMusicSound(vehicle));
            }
        };

        VehicleEntity.fireSound = vehicle -> Minecraft.getInstance().getSoundManager().play(new VehicleFireSoundInstance.VehicleFireSound(vehicle));

        FastThrowableProjectile.flySound = entity -> Minecraft.getInstance().getSoundManager().play(new FastProjectileSoundInstance.FlySound(entity));
        FastThrowableProjectile.nearFlySound = entity -> Minecraft.getInstance().getSoundManager().play(new FastProjectileSoundInstance.NearFlySound(entity));
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/init/ModSoundInstances.kt
package com.atsuishio.superbwarfare.init

import com.atsuishio.superbwarfare.client.sound.FastProjectileSoundInstance
import com.atsuishio.superbwarfare.client.sound.HornSoundInstance
import com.atsuishio.superbwarfare.client.sound.StagedVehicleEngineSound
import com.atsuishio.superbwarfare.client.sound.SteelCoilMoveSoundInstance
import com.atsuishio.superbwarfare.client.sound.VehicleFireSoundInstance
import com.atsuishio.superbwarfare.client.sound.VehicleSoundInstance
import com.atsuishio.superbwarfare.entity.living.SteelCoilEntity
import com.atsuishio.superbwarfare.entity.projectile.FastThrowableProjectile
import com.atsuishio.superbwarfare.entity.vehicle.base.VehicleEntity
import com.atsuishio.superbwarfare.tools.mc
import net.neoforged.api.distmarker.Dist
import net.neoforged.api.distmarker.OnlyIn
import java.util.function.Consumer

@OnlyIn(Dist.CLIENT)
object ModSoundInstances {
    fun init() {
        VehicleEntity.playTrackSound =
            Consumer { mc.soundManager.play(VehicleSoundInstance.TrackSound(it)) }
        VehicleEntity.playEngineSound =
            Consumer { vehicle ->
                vehicle ?: return@Consumer
                if (!StagedVehicleEngineSound.play(vehicle)) {
                    mc.soundManager.play(VehicleSoundInstance.EngineSound(vehicle))
                }
            }
        VehicleEntity.playEngineStartSound =
            Consumer { vehicle -> vehicle?.let(StagedVehicleEngineSound::playStart) }
        VehicleEntity.playEngineStopSound =
            Consumer { vehicle -> vehicle?.let(StagedVehicleEngineSound::playStop) }
        VehicleEntity.playSwimSound =
            Consumer { mc.soundManager.play(VehicleSoundInstance.SwimSound(it)) }
        VehicleEntity.playHornSound =
            Consumer { mc.soundManager.play(HornSoundInstance.VehicleHornSound(it)) }
        VehicleEntity.playStukaSound =
            Consumer { mc.soundManager.play(VehicleSoundInstance.StukaSound(it)) }
        VehicleEntity.playHeliCrashSound =
            Consumer { mc.soundManager.play(VehicleSoundInstance.HeliCrashSound(it)) }
        VehicleEntity.playVehicleSkipSound =
            Consumer { mc.soundManager.play(VehicleSoundInstance.SkipSound(it)) }
//        VehicleEntity.playInCarMusic = vehicle -> {
//            if (NetMusicCompatHolder.canPlayMusic(vehicle)) {
//                NetMusicCompatHolder.playMusic(vehicle);
//            } else {
//                mc.soundManager.play(new InCarMusicInstance.InCarMusicSound(vehicle));
//            }
//        };
        VehicleEntity.playFireSound =
            Consumer(VehicleFireSoundInstance.VehicleFireSound::play)
        FastThrowableProjectile.playFlySound =
            Consumer { mc.soundManager.play(FastProjectileSoundInstance.FlySound(it)) }
        SteelCoilEntity.playMoveSound =
            Consumer { mc.soundManager.play(SteelCoilMoveSoundInstance.SteelCoilMoveSound(it)) }
    }
}
